infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/external/ExternalServerExposerProvider.java

```diff
@@ -28,16 +33,49 @@
 @Singleton
 public class ExternalServerExposerProvider<T extends KubernetesEnvironment>
     extends AbstractExposureStrategyAwareProvider<ExternalServerExposer<T>> {
+
+  private final boolean forceSubdomainEndpoints;
+  private final ExternalServerExposer<T> combinedInstance;
+
   @Inject
   public ExternalServerExposerProvider(
       @Named("che.infra.kubernetes.server_strategy") String exposureStrategy,
       @Named("che.infra.kubernetes.singlehost.workspace.exposure") String exposureType,
+      @Named("che.infra.kubernetes.singlehost.workspace.expose_devfile_endpoints_on_subdomains")
+          boolean exposeDevfileEndpointsOnSubdomains,
+      @Named(INGRESS_DOMAIN_PROPERTY) String domain,
+      @Named("infra.kubernetes.ingress.annotations") Map<String, String> annotations,
+      @Nullable @Named("che.infra.kubernetes.ingress.labels") String labelsProperty,
+      @Nullable @Named("che.infra.kubernetes.ingress.path_transform") String pathTransformFmt,
       Map<WorkspaceExposureType, ExternalServerExposer<T>> exposers) {
 
     super(
         exposureStrategy,
         exposureType,
         exposers,
         "Could not find an external server exposer implementation for the exposure type '%s'.");
+    if (SINGLE_HOST_STRATEGY.equals(exposureStrategy) && exposeDevfileEndpointsOnSubdomains) {
+      this.forceSubdomainEndpoints = true;
+      this.combinedInstance =
+          new CombinedSingleHostServerExposer<>(
+              new IngressServerExposer<>(
```

Comment:
If you know how to satisfy constructor with 2 same parameters in guice, I'm in https://github.com/eclipse/che/pull/17928/files#diff-18119655c32781f07ab80eee19f8e632R32

Reply:
I've updated the PR so now I'm injecting the "multi-host" exposer here.

Still creating the `CombinedSingleHostServerExposer` manually. I don't know how to easily bind this one, because I need to return Combined from this Provider, but I need this Provider to construct the Combined. So I would probably somehow need yet another provider that will wire up everything together. I'm not sure it's worth. WDYT @metlos ?